---
title: "Base workflow of flowSpy in use case 1 and 2"
author: "Yuting Dai"
package: flowSpy
date: "`r Sys.Date()`"
output:
  prettydoc::html_pretty:
    toc: true
    theme: cayman
    highlight: github
  pdf_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Base_workflow}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
  %\VignetteEncoding{UTF-8}
---

```{r use-libs, echo = FALSE, message = FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = FALSE, 
                      warning = FALSE, message = TRUE)
```


## Introduction

To validate the cellular subpopulations identified by flowSpy, we used a 13-marker panel mass cytometry dataset of healthy human bone marrow. This dataset was generated from Bendall et al [1] and completed quality control by Herring et al [2], which could be downloaded from FlowRepository database [3] (https://flowrepository.org/id/FR-FCM-ZY9R). The aim of this use case was to identify the cellular subpopulations and construct a tree-shaped trajectory, which could reveal the human hematopoietic differentiation hierarchy.

This tutorial contains key steps of `flowSpy` base workflow, including how to build an FSPY object, how to run clustering and dimensionality reduction, how to build a tree based on [minimum spanning tree](https://en.wikipedia.org/wiki/Minimum_spanning_tree) (MST) algorithm, how to run pseudotime and how to identify intermediate state cells. 

## Workflow

``` {r quick-build, eval = TRUE, fig.width = 6, fig.height = 5}

# Loading packages
suppressMessages({
library(ggplot2)
library(flowCore)
library(pheatmap)
library(flowSpy)
library(stringr)
})

#########################
# Read Mass Cytometry Data
# It can be downloaded via `git clone https://github.com/ytdai/flowSpy-dataset.git` 
# fcs.path musted be modified based on the download directory from GitHub
fcs.path <- "../../flowSpy-dataset/FCS/usecase1_2/"
fcs.file <- paste0(fcs.path, "FR-FCM-ZY9R-Bone_Marrow_cytof.fcs")

###########################################
#   Get the expression matrix from FCS file
###########################################

# Solution 1
# Read FCS data via flowCore::read.FCS
# Expression data matrix from this method need to
# be performed compensation adjustment and transformation
# manually using flowCore
cytof.data <- flowCore::read.FCS(filename = fcs.file)
# show elements in mass cytometry data
cytof.data
# fetching expression data 
exp.data <- cytof.data@exprs

# Solution 2
# Read FCS data via flowSpy::runExprsExtract
# ** This solution is recommended
# ** Use case 1 and 2 follow this solution
exp.data <- runExprsExtract(fcs.file, showDesc = FALSE, transformMethod = "autoLgcl")

# Fetching CD markers
markers <- colnames(exp.data)
markers.idx <- match(markers, colnames(exp.data))

# Build an FSPY object
# If you don't want to see the running log information, set verbose FALSE
# If there is only one case in your analysis workflow, you can just set stage "D0"
meta.data <- data.frame(cell = rownames(exp.data),
                        stage = "D0")
fspy <- createFSPY(raw.data = exp.data, markers = markers,
                   meta.data = meta.data,
                   normalization.method = "none")

# Cluster cells by SOM algorithm
# Set random seed to make results reproducible
set.seed(6)
fspy <- runCluster(fspy, cluster.method = "som", xdim = 10, ydim = 10)

# Cluster based downsampling
# The total cell number is 236,187, and we can just keep 10% cells to reduce 
# computation load and improve computation time.
# Downsampling by setting downsampleing.size 0.1
set.seed(1)
fspy <- processingCluster(fspy, perplexity = 5,downsampling.size = 0.1)

# Now only 23,664 cells are enrolled in the dimensionality reduction
fspy

# run Principal Component Analysis (PCA)
fspy <- runFastPCA(fspy)

# run t-Distributed Stochastic Neighbor Embedding (tSNE)
set.seed(1)
fspy <- runTSNE(fspy, dims = 2)

# run Diffusion map
fspy <- runDiffusionMap(fspy)

# run Uniform Manifold Approximation and Projection (UMAP)
fspy <- runUMAP(fspy)

###########################################
# This is visualization module
###########################################

# Plot 2D PCA. And cells are colored by CD3 expression
plot2D(fspy, item.use = c("PC_1", "PC_2"), color.by = "CD3", 
       alpha = 1, main = "PCA", category = "numeric") + 
  scale_colour_gradientn(colors = c("#00599F","#EEEEEE","#FF3222"))

# Plot 2D tSNE. And cells are colored by CD3 expression
plot2D(fspy, item.use = c("tSNE_1", "tSNE_2"), color.by = "CD3", 
       alpha = 1, main = "tSNE", category = "numeric") + 
  scale_colour_gradientn(colors = c("#00599F","#EEEEEE","#FF3222"))

# Plot 2D diffusion maps. And cells are colored by CD3 expression
plot2D(fspy, item.use = c("DC_1", "DC_2"), color.by = "CD3", 
       alpha = 1, main = "diffusion maps", category = "numeric") + 
  scale_colour_gradientn(colors = c("#00599F","#EEEEEE","#FF3222"))

# Plot 2D UMAP. And cells are colored by CD3 expression
plot2D(fspy, item.use = c("UMAP_1", "UMAP_2"), color.by = "CD3", 
       alpha = 1, main = "UMAP", category = "numeric") + 
  scale_colour_gradientn(colors = c("#00599F","#EEEEEE","#FF3222"))

# Plot 2D tSNE. And cells are colored by cluster id
plot2D(fspy, item.use = c("tSNE_1", "tSNE_2"), color.by = "cluster.id", 
       alpha = 1, main = "tSNE", category = "categorical", show.cluser.id = T)

# Plot 2D UMAP. And cells are colored by cluster id
plot2D(fspy, item.use = c("UMAP_1", "UMAP_2"), color.by = "cluster.id", 
       alpha = 1, main = "UMAP", category = "categorical", show.cluser.id = T)

# Plot 3D UMAP. And cells are colored by CD45RA markers expression
plot3D(fspy, item.use = c("DC_1", "DC_2", "DC_3"), color.by = "CD3", 
       main = "diffusion maps CD3", category = "numeric", size = 0.2, 
       color.theme = c("#00599F","#EEEEEE","#FF3222"))

###########################################
#   Trajectory 
###########################################

# flowSpy provides five method to build the tree-shaped trajectory: 
# 1. Raw expression matrix
# 2. PCA
# 3. tSNE
# 4. Diffusion maps
# 5. UMAP

# 1. Raw expression matrix
fspy <- buildTree(fspy, dim.type = "raw")
# Tree plot
plotTree(fspy, color.by = "CD3", show.node.name = T, cex.size = 1) + 
  scale_colour_gradientn(colors = c("#00599F", "#EEEEEE", "#FF3222"))

# 2. PCA
fspy <- buildTree(fspy, dim.type = "pca", dim.use = 1:4)
# Tree plot
plotTree(fspy, color.by = "CD3", show.node.name = T, cex.size = 1) + 
  scale_colour_gradientn(colors = c("#00599F", "#EEEEEE", "#FF3222"))

# 3. tSNE
fspy <- buildTree(fspy, dim.type = "tsne", dim.use = 1:2)
# Tree plot
plotTree(fspy, color.by = "CD3", show.node.name = T, cex.size = 1) + 
  scale_colour_gradientn(colors = c("#00599F", "#EEEEEE", "#FF3222"))

# 4. Diffusion maps
fspy <- buildTree(fspy, dim.type = "dc", dim.use = 1:3)
# Tree plot
plotTree(fspy, color.by = "CD3", show.node.name = T, cex.size = 1) + 
  scale_colour_gradientn(colors = c("#00599F", "#EEEEEE", "#FF3222"))

# 5. UMAP
fspy <- buildTree(fspy, dim.type = "umap", dim.use = 1:2)
# Tree plot
plotTree(fspy, color.by = "CD3", show.node.name = T, cex.size = 1) + 
  scale_colour_gradientn(colors = c("#00599F", "#EEEEEE", "#FF3222"))

# The topology of a trajectory is mainly based on the interrelation 
# of cell clusters, coordinates and dimensions, and in use case 1 and 
# 2, we use "tsne" to construct the trajectory
fspy <- buildTree(fspy, dim.type = "tsne", dim.use = 1:2)
# Tree plot
plotTree(fspy, color.by = "CD3", show.node.name = T, cex.size = 1) + 
  scale_colour_gradientn(colors = c("#00599F", "#EEEEEE", "#FF3222"))

plotTree(fspy, color.by = "CD34", show.node.name = T, cex.size = 1) + 
  scale_colour_gradientn(colors = c("#00599F", "#EEEEEE", "#FF3222"))


############# Modify branch id
fspy@meta.data$branch.id[fspy@meta.data$cluster.id %in% c(9,25,90,30)] = 12
fspy@meta.data$branch.id[fspy@meta.data$cluster.id %in% c(5,62)] = 6
fspy@meta.data$branch.id[fspy@meta.data$cluster.id %in% c(38,87)] = 3

fspy@meta.data$branch.id[fspy@meta.data$branch.id %in% c(12)] = "HSCs"
fspy@meta.data$branch.id[fspy@meta.data$branch.id %in% c(1,2)] = "CD8 T cells"
fspy@meta.data$branch.id[fspy@meta.data$branch.id %in% c(4,5)] = "CD4 T cells"
fspy@meta.data$branch.id[fspy@meta.data$branch.id %in% c(3,9,11)] = "Megakaryocytic"
fspy@meta.data$branch.id[fspy@meta.data$branch.id %in% c(6)] = "NK cells"
fspy@meta.data$branch.id[fspy@meta.data$branch.id %in% c(7)] = "B cells"
fspy@meta.data$branch.id[fspy@meta.data$branch.id %in% c(8,10)] = "Myeloid"

# Run differential expressed markers of different branch
diff.info <- runDiff(fspy)
head(diff.info)

# plot tree
plotTree(fspy, color.by = "branch.id", show.node.name = T, cex.size = 1) 

plotCluster(fspy, item.use = c("tSNE_1", "tSNE_2"), category = "categorical",
            size = 100, color.by = "branch.id", show.cluser.id = T) 


# Plot 2D and colored by branch id
plot2D(fspy, item.use = c("tSNE_1", "tSNE_2"), color.by = "branch.id", 
       alpha = 1, main = "tSNE", category = "categorical", show.cluser.id = F)

plot2D(fspy, item.use = c("PC_1", "PC_2"), color.by = "branch.id", 
       alpha = 1, main = "PCA", category = "categorical", show.cluser.id = F)

plot2D(fspy, item.use = c("DC_1", "DC_2"), color.by = "branch.id", 
       alpha = 1, main = "diffusion maps", category = "categorical", show.cluser.id = F)

plot2D(fspy, item.use = c("UMAP_1", "UMAP_2"), color.by = "branch.id", 
       alpha = 1, main = "UMAP", category = "categorical", show.cluser.id = F)

# plot for clusters 
plotCluster(fspy, item.use = c("PC_1", "PC_2"), category = "numeric",
            size = 10, color.by = "CD3", main = "PCA of clusters") + 
  scale_colour_gradientn(colors = c("#00599F", "#EEEEEE", "#FF3222"))

plotCluster(fspy, item.use = c("tSNE_1", "tSNE_2"), category = "numeric",
            size = 100, color.by = "CD3", main = "tSNE of clusters") + 
  scale_colour_gradientn(colors = c("#00599F", "#EEEEEE", "#FF3222"))

plotCluster(fspy, item.use = c("tSNE_1", "tSNE_2"), category = "categorical",
            size = 100, color.by = "branch.id", show.cluser.id = T)

# plot heatmap of clusters and branches
plotClusterHeatmap(fspy)
plotClusterHeatmap(fspy, color = colorRampPalette(c("purple","white","yellow"))(100))

plotBranchHeatmap(fspy, clustering_method = "ward.D")

###########################################
#   Pseudotime 
###########################################

# Set HSPCs as root cells
fspy <- defRootCells(fspy, root.cells = c(9,25,90,30))
fspy <- runPseudotime(fspy, verbose = T, dim.type = "raw")

###### Intermediate state cells for CD8 T cells
fspy <- defLeafCells(fspy, leaf.cells = c(23,28,51))
fspy <- runWalk(fspy, backward.walk = F, verbose = T)
fspy@meta.data$traj.value.log.CD8T <- fspy@meta.data$traj.value.log

###### Intermediate state cells for CD4 T cells
fspy <- defLeafCells(fspy, leaf.cells = c(52,78,46))
fspy <- runWalk(fspy, backward.walk = F, verbose = T)
fspy@meta.data$traj.value.log.CD4T <- fspy@meta.data$traj.value.log

###### Intermediate state cells for NK cells
fspy <- defLeafCells(fspy, leaf.cells = c(54,48))
fspy <- runWalk(fspy, backward.walk = F)
fspy@meta.data$traj.value.log.NK <- fspy@meta.data$traj.value.log

###### Intermediate state cells for B cells
fspy <- defLeafCells(fspy, leaf.cells = c(11,89,94))
fspy <- runWalk(fspy, backward.walk = F)
fspy@meta.data$traj.value.log.B <- fspy@meta.data$traj.value.log

###### Intermediate state cells for monocytes and granulocytes
fspy <- defLeafCells(fspy, leaf.cells = c(49,93,42))
fspy <- runWalk(fspy, backward.walk = F)
fspy@meta.data$traj.value.log.MY <- fspy@meta.data$traj.value.log

###### Intermediate state cells for megakaryocyte and erythrocyte
fspy <- defLeafCells(fspy, leaf.cells = c(14,31,100,63,98))
fspy <- runWalk(fspy, backward.walk = F)
fspy@meta.data$traj.value.log.ME <- fspy@meta.data$traj.value.log

# Plot 2D tSNE. 
fspy@meta.data$stage <- fspy@meta.data$branch.id

plot2D(fspy, item.use = c("tSNE_1", "tSNE_2"), category = "numeric",
            size = 1, color.by = "pseudotime") + 
 scale_colour_gradientn(colors = c("#F4D31D", "#FF3222","#7A06A0"))

# Tree plot
plotTree(fspy, color.by = "pseudotime", cex.size = 1) + 
 scale_colour_gradientn(colors = c("#F4D31D","#FF3222","#7A06A0"))

# pseudotime density
plotPseudotimeDensity(fspy, adjust = 2)

plotPseudotimeTraj(fspy, var.cols = T) + 
 scale_colour_gradientn(colors = c("#F4D31D", "#FF3222","#7A06A0"))

### fetch plot information
plot.meta <- fetchPlotMeta(fspy, markers = markers)

# heatmap for CD8 T cells
plot.meta.sub <- plot.meta[which(plot.meta$traj.value.log.CD8T > 0), ]
plot.meta.sub <- plot.meta.sub[order(plot.meta.sub$pseudotime), ]
pheatmap(t(plot.meta.sub[, markers]), scale  = "row",
         cluster_rows = T, cluster_cols = F, cluster_method = "ward.D",
         color = colorRampPalette(c("blue","blue","white","red","red"))(100),
         fontsize_col = 0.01)

# heatmap for CD4 T cells
plot.meta.sub <- plot.meta[which(plot.meta$traj.value.log.CD4T > 0), ]
plot.meta.sub <- plot.meta.sub[order(plot.meta.sub$pseudotime), ]
pheatmap(t(plot.meta.sub[, markers]), scale  = "row",
         cluster_rows = T, cluster_cols = F, cluster_method = "ward.D",
         color = colorRampPalette(c("blue","blue","white","red","red"))(100),
         fontsize_col = 0.01)

# heatmap for NK cells
plot.meta.sub <- plot.meta[which(plot.meta$traj.value.log.NK > 0), ]
plot.meta.sub <- plot.meta.sub[order(plot.meta.sub$pseudotime), ]
pheatmap(t(plot.meta.sub[, markers]), scale  = "row",
         cluster_rows = T, cluster_cols = F, cluster_method = "ward.D",
         color = colorRampPalette(c("blue","blue","white","red","red"))(100),
         fontsize_col = 0.01)

# heatmap for B cells
plot.meta.sub <- plot.meta[which(plot.meta$traj.value.log.B > 0), ]
plot.meta.sub <- plot.meta.sub[order(plot.meta.sub$pseudotime), ]
pheatmap(t(plot.meta.sub[, markers]), scale  = "row",
         cluster_rows = T, cluster_cols = F, cluster_method = "ward.D",
         color = colorRampPalette(c("blue","blue","white","red","red"))(100),
         fontsize_col = 0.01)

# heatmap for monocytes and granulocytes
plot.meta.sub <- plot.meta[which(plot.meta$traj.value.log.MY > 0), ]
plot.meta.sub <- plot.meta.sub[order(plot.meta.sub$pseudotime), ]
pheatmap(t(plot.meta.sub[, markers]), scale  = "row",
         cluster_rows = T, cluster_cols = F, cluster_method = "ward.D",
         color = colorRampPalette(c("blue","blue","white","red","red"))(100),
         fontsize_col = 0.01)

# heatmap for megakaryocyte and erythrocyte
plot.meta.sub <- plot.meta[which(plot.meta$traj.value.log.ME > 0), ]
plot.meta.sub <- plot.meta.sub[order(plot.meta.sub$pseudotime), ]
pheatmap(t(plot.meta.sub[, markers]), scale  = "row",
         cluster_rows = T, cluster_cols = F, cluster_method = "ward.D",
         color = colorRampPalette(c("blue","blue","white","red","red"))(100),
         fontsize_col = 0.01)


```

## Session information

```{r sess-info}
sessionInfo()
```

## References

1.	Bendall SC, Simonds EF, Qiu P, Amir el AD, Krutzik PO, Finck R, Bruggner RV, Melamed R, Trejo A, Ornatsky OI, et al: Single-cell mass cytometry of differential immune and drug responses across a human hematopoietic continuum. Science 2011, 332:687-696.

2.	Herring CA, Banerjee A, McKinley ET, Simmons AJ, Ping J, Roland JT, Franklin JL, Liu Q, Gerdes MJ, Coffey RJ, Lau KS: Unsupervised Trajectory Analysis of Single-Cell RNA-Seq and Imaging Data Reveals Alternative Tuft Cell Origins in the Gut. Cell Syst 2018, 6:37-51 e39.

3.	Spidlen J, Breuer K, Rosenberg C, Kotecha N, Brinkman RR: FlowRepository: a resource of annotated flow cytometry datasets associated with peer-reviewed publications. Cytometry A 2012, 81:727-731.









